### DISCUSSION TITLE
When Effect cancelled, it can still send actions

### DISCUSSION THREAD

**Discussion by BeetLab** [pointfreeco/swift-composable-architecture#3488] [Q&A]
Hi! I've met an unexpected (for me) behaviour of Effect. 

I have timeout effect, that is cancelled when main peace of work finishes. But even after cancellation of `Tokens.timeout` the Task inside can still send action and it would be consumed by the corresponding reducer. Why is send still usable after Effect cancellation? 
I understand, that Task inside Effect is top-level and doesn't know about Effect's task cancellation, but why the `send`  can still be usable? 
Thanks!

```
                let timeoutTask = Effect<Action>.run { send in
                    Task {
                        try? await Task.sleep(seconds: 6)
                        await send(.delegate(.timeoutExceeded))
                    }
                }
                .cancellable(id: Tokens.timeout)

```

**Comment by mbrandonw** [comment]
Hi @BeetLab, I don't believe this is actually possible. First, based on the code you shared above, the `timeoutTask` effects completes almost immediately since nothing is `await`ing inside the effect and you are instead performing all work inside a `Task`. Typically you would just perform that work right in the effect and not spin up a `Task`. Also, when the `send(.delegate(.timeoutExceeded))` line executes it should even emit a runtime warning letting you know that you are trying to send actions after the effect has completed, which isn't allowed.

And further, effects are completed when they are canceled and so should not be able to emit any data from a cancelled effect.

If you believe you are seeing otherwise then can you provide a minimal project that reproduces the problem? Ideally this would be done as a unit test that fails that you believe should pass. And please try to include only the essential details and remove as many superfluous parts as possible.

**Comment by BeetLab** [comment]
Well, you are right. I didn't manage to reproduce the issue in unit tests, but looks like there's something in out project's setup that changes the behaviour (will check).

Just in case - here's a test with setup I've had in out project (and no runtime warnings after completing the effect - maybe because of test store using?)
[CancelledEffect.swift.zip](https://github.com/user-attachments/files/17729145/CancelledEffect.swift.zip)
